Meals remaining

elijahverdoorn opened the pull request:
Ok. I've got this working. The code's not beautiful, but it does work. I made a few decisions in there that we might want to discuss:

* Layout of settings
* Redirection to settings from SIS
* Re-naming variables in the SIS

I'd appreciate suggestions and tweaks.
(I'm especially not proud of the error checking code, there's got to be a better way to do that)

Closes #303

elijahverdoorn committed:
Add fetching for the meals remaining

Weekly and daily meals remaining were never being tracked, this adds that functionaily. We're not storing the username and password properly though, so this doesn't fully work (yet).

elijahverdoorn committed:
Error checking in fetching remaining meals data

elijahverdoorn committed:
Re-add St. Olaf username and password fields to the settings menu

Using the same components that we removed when we added OAuth.

elijahverdoorn committed:
Error checking around login process

Add error checking aroud null strings in the username and password fields

elijahverdoorn committed:
Add redirection to the settings instead of the sis login view

Since we now want the user to give us two pieces of information, let's just use the settings menu to get that info. The button even says that it goes to settings.

zimmer18 commented on lib/financials.js:
```diff
@@ -104,17 +104,30 @@ async function getFinancialDataFromServer(): Promise<FinancialDataShapeType> {
   ])
 
-  return {flex, ole, print}
+  return {flex, ole, print};
```
; ?

hawkrives replied:
Bad semicolon :)
And good eye, @zimmer18.

drewvolz commented on views/sis/balances.js:
```diff
@@ -71,6 +75,7 @@ export default class BalancesView extends React.Component {
   }
 
   loadIfLoggedIn = async () => {
+    console.log('TESTING')
```
:zap:

hawkrives replied:
Remove this `log`, please

drewvolz commented on lib/financials.js:
```diff
@@ -104,17 +104,30 @@ async function getFinancialDataFromServer(): Promise<FinancialDataShapeType> {
+    })
+    return({"weeklyMeals": values[4], "dailyMeals": values[3]})
+  } catch (error) {
+    console.log("error in fetching meal data")
```
Maybe we should present an alert, or some information for the user to know that it could not be fetched?

hawkrives replied:
Nah. It'll just say N/A in the view.

drewvolz replied:
I really think being clear is important here. Otherwise, you get 1-star reviews on the store saying it does not work.

hawkrives replied:
I'm just not sure what someone would expect. We have no idea what the failure cases would be, really, so we can't provide tailored error messages. The most common case will be that your password has changed, but that should be dealt with similarly to the SIS data, I think.

drewvolz commented on views/settings/index.js:
```diff
@@ -36,50 +43,81 @@ export default class SettingsView extends React.Component {
+    this.setState({loadingGoogle: true})
     CookieManager.clearAll((err, res) => {
       if (err) {
         console.log(err)
```
Maybe we could let the user know we could not log them out?

hawkrives commented on lib/financials.js:
```diff
@@ -104,17 +104,30 @@ async function getFinancialDataFromServer(): Promise<FinancialDataShapeType> {
+    let values = []
+    data.map((item) => {
+      values.push(item.next.next.children[0].data)
+    })
```
Please don't use `map` for side-effects; `map` should be a pure function, not mutating anything outside of its scope.

Instead, you could do 

``` js
let values = data.map((item) => {
  return item.next.next.children[0].data
})
```

hawkrives commented on lib/login.js:
```diff
@@ -28,5 +29,43 @@ export function clearLoginCredentials(): Promise<any> {
+
+    return {result}
+  } else {
+    return {false}
```
According to your type defs, this should be `return {result: false}`

hawkrives commented on views/settings/index.js:
```diff
@@ -36,50 +43,81 @@ export default class SettingsView extends React.Component {
       if (err) {
         console.log(err)
       }
       console.log(res)
```
Remove this `log`

hawkrives commented on views/settings/index.js:
```diff
@@ -36,50 +43,81 @@ export default class SettingsView extends React.Component {
+    this.setState({loadingStOlaf: true})
+    let {username, password} = this.state
+    if (username) {
+      username.replace('@stolaf.edu', '') // just in case someone uses their full email, throw the part we don't need away
```
Might as well just make this generic; remove anything after `@`? Something like `.replace(/@.*$/, '')`?

hawkrives commented on views/settings/index.js:
```diff
@@ -102,36 +140,106 @@ export default class SettingsView extends React.Component {
+          {usernameCell}
+          {passwordCell}
+          {loginButton}
           {actionCell}
```
I don't like having both of the login buttons right next to each other. Can we pull the "Sign In with Google" button into its own `<Section/>`?

hawkrives commented on views/sis/balances.js:
```diff
@@ -59,6 +61,8 @@ export default class BalancesView extends React.Component {
     ole: null|number,
     print: null|number,
+    weeklyMeals: null|number,
+    dailyMeals: null|number,
```
Yay flow types!

hawkrives commented on views/sis/balances.js:
```diff
@@ -92,6 +97,13 @@ export default class BalancesView extends React.Component {
         let {flex, ole, print} = data
         this.setState({flex, ole, print})
       }
+      data = await getWeeklyMealsRemaining()
```
Maybe don't reuse the variable? No harm, I suppose, but I dislike reusing variables in the same scope…

hawkrives commented on views/sis/balances.js:
```diff
@@ -174,12 +186,12 @@ export default class BalancesView extends React.Component {
             <Cell cellStyle='RightDetail'
               title='Daily Meals Left'
-              detail='Not Available'
+              detail = {dailyMeals}
```
No spaces around the `=`

hawkrives commented on views/sis/balances.js:
```diff
@@ -174,12 +186,12 @@ export default class BalancesView extends React.Component {
             <Cell cellStyle='RightDetail'
               title='Weekly Meals Left'
-              detail='Not Available'
+              detail = {weeklyMeals}
```
No spaces around the `=`

hawkrives commented on views/settings/index.js:
```diff
@@ -102,36 +140,106 @@ export default class SettingsView extends React.Component {
     let accountSection = (
       <View>
-        <Section>
+        <Section header='LOGIN' footer='Note: This application requires logging in to Google and inputting your St. Olaf username and password in order use all features.'>
```
I feel like this should say `[…] requires both logging into Google and […]`, with a `both` in there somewhere?

drewvolz replied:
"in order **to** use all features"

elijahverdoorn committed:
Address PR feedback on Meals addition to SIS

hawkrives commented on views/settings/index.js:
```diff
@@ -102,36 +140,106 @@ export default class SettingsView extends React.Component {
+          value={username}
+          returnKeyType='next'
+          onChangeText={text => this.setState({username: text})}
+          onSubmitEditing={this.focusPassword}
```
`this.focusPassword` needs to be restored from the history

hawkrives replied:
done

hawkrives commented on views/settings/index.js:
```diff
@@ -102,36 +140,106 @@ export default class SettingsView extends React.Component {
+
+    let usernameCell = (
+      <CustomCell contentContainerStyle={styles.loginCell}>
+        <Text onPress={this.focusUsername} style={styles.label}>Username</Text>
```
`this.focusUsername` needs to be restored from the history

elijahverdoorn replied:
Gotcha. I've run out of time for the moment, but I'll be back on it later. Good catch.

hawkrives replied:
done

hawkrives commented on views/settings/index.js:
```diff
@@ -102,36 +140,106 @@ export default class SettingsView extends React.Component {
+
+    let passwordCell = (
+      <CustomCell contentContainerStyle={styles.loginCell}>
+        <Text onPress={this.focusPassword} style={styles.label}>Password</Text>
```
`this.focusPassword` needs to be restored from the history

hawkrives replied:
done

hawkrives reviewed (changes requested):
These functions

hawkrives committed:
Re-add more of the meal fetching stuff

Re-added some of the fetching/login checking from before OAuth for the purpose of checking credientials for the meals remaining.

hawkrives committed:
Fix the login button

also add some idiot proofing around the username

hawkrives committed:
default username/pw to empty strings

hawkrives committed:
declare the component refs

hawkrives committed:
fix eslint warning

hawkrives committed:
re-add focus{Username,Password} methods

hawkrives committed:
address last part of my own review

hawkrives committed:
move placeholderTextColor into colors.js

hawkrives committed:
break login buttons into two sections

hawkrives committed:
fix typo that was toggling the wrong login button

hawkrives committed:
fix lint warnings

hawkrives committed:
parallelize some data fetching

hawkrives committed:
return N/A for meal plan balances when no value found

hawkrives committed:
remove logout of both credentials on the "st olaf logout" button

hawkrives committed:
clarify requirements on sis/error-screen

hawkrives committed:
check sis login on view load

hawkrives committed:
simplify credentials:valid removal?

hawkrives committed:
update loggedInGoogle when login happens

hawkrives committed:
update loggedInGoogle when logging out
